drivers/source/usb/USBDevice.cpp

```diff
@@ -1771,3 +1776,17 @@ void USBDevice::_run_later(void (USBDevice::*function)())
 {
     _post_process = function;
 }
+
+void USBDevice::enable_phy(bool enabled)
+{
+    core_util_critical_section_enter();
+    if (_phy_enabled != enabled) {
+        if (enabled) {
+            USBDevice::connect();
```

Comment:
So it seems that `deinit` calls `disconnect` and `connect` calls `init` so we should be good there. But i'm not 100% sure it will restore the state. Is calling `disconnect` and `phy->deinit` legal in every state? Basically if I call `enable_phy(false)`, go to sleep, wake up, call `enable_phy(true)` and try to continue using USB will it work?

Reply:
I programmed a K64F with a firmware that allowed me to call `enable_phy()` by pressing a button to enable/disable USB phy.
After `enable_phy(false)`, nothing can be received on the USB interface and the device goes to deep sleep. After `enable_phy(true)`, bytes can once again be received on the USB interface.
Is that what youwere not sure about?

Reply:
Yeah, the test you performed is a good one, but I'm not sure it's definitive for all implementation based on the specification. Could you check that it's actually mandated by the spec? Russ would be able to help I'm sure.

Reply:
Hello @c1728p9 ,
Could you please comment?

Reply:
Unlike some of the other peripherals, the USB code to disable deep sleep is target specific and in a subclass of USBPhy. There is no code common to all targets which locks deep sleep. Depending on implementation, targets could unlock deep sleep on a call to disconnect(), or even do it automatically when USB power is lost. Currently all the implementation I'm aware of only unlock deep sleep in deinit() though.

> Is calling disconnect and phy->deinit legal in every state?

Yes, this is equivalent to the user unplugging the device. This may have negative side effects though, such as corrupting a mass storage drive if disconnecting in the middle of a write.

> But i'm not 100% sure it will restore the state.

The call to enable_phy(true) can make the device re-enumerate, but can't directly restore the state. The USB host controls the enumeration process so it chooses when and how the device will be restored. Even if the device state is returned to what it was before the disconnect, the host PC software may not be where it left off - for example a serial port may need to be re-opended by the user.

The safest thing to do would be only deinit() when USB power is lost and init() and connect() when it is detected. That way there isn't any state to restore.

Finally, one thing to note is that suspend/resume is already terminology used by the USB standard. It referrs to a USB host initiated low power mode. This may add some confusion to users. I'm not sure of a good way to avoid this though.

Reply:
If it's also not guaranteed to be clean operation in terms of state being preserved I would say suspend and resume doesn't necessarily work. Is directly calling init/deinit or connect/disconnect a viable solution? Documentation can be clear that it's up to the user to make sure all the transfers have concluded?

Reply:
[Done](https://github.com/ARMmbed/mbed-os/pull/11635#issuecomment-540425707).